File: binV7/third_problem.rb
require 'json'

n = ARGV[1].to_i
n = 2 if n.zero?

products = [
  { "name": "Apples", "quantity_sold": 120 },
  { "name": "Oranges", "quantity_sold": 75 },
  { "name": "Bananas", "quantity_sold": 200 }
]

sorted_products = products.sort_by { |product| product[:quantity_sold] }.reverse

most_sold_products = sorted_products.take(n)

print JSON.pretty_generate({ data: most_sold_products })

File: quizzes8/third_problem.ts
export {};

type Product = {
  name: string;
  quantity_sold: number;
};

// @ts-ignore
const n: number = Number(process.argv[3]) || 2;

const products: Product[] = [
  { name: "Apples", quantity_sold: 120 },
  { name: "Oranges", quantity_sold: 75 },
  { name: "Bananas", quantity_sold: 200 },
];

const findTopSoldProductsWithSort = (): Product[] => {
  return products
    .toSorted((a, b) => b.quantity_sold - a.quantity_sold)
    .slice(0, n);
};

const findTopSoldProductsWithLoop = (): Product[] => {
  const topSoldProducts: Product[] = [];

  for (let i = 0; i < products.length; i++) {
    const currentProduct: Product = products[i];

    if (topSoldProducts.length < n) {
      topSoldProducts.push(currentProduct);
      topSoldProducts.sort((a, b) => b.quantity_sold - a.quantity_sold);
    } else if (
      currentProduct.quantity_sold > topSoldProducts[n - 1].quantity_sold
    ) {
      topSoldProducts[n - 1] = currentProduct;
      topSoldProducts.sort((a, b) => b.quantity_sold - a.quantity_sold);
    }
  }

  return topSoldProducts;
};

console.log(JSON.stringify({ data: findTopSoldProductsWithSort() }, null, 2));
console.log(JSON.stringify({ data: findTopSoldProductsWithLoop() }, null, 2));
